app/wallet/oauth.go

```diff
@@ -0,0 +1,199 @@
+
+	if cachedUser := currentCache.get(userInfo.Sub); cachedUser != nil {
+		log.Debugf("user found in cache")
+		return cachedUser, nil
+	}
+
+	//Check if we have the user by IDP ID first
+	user, err := GetDBUserG(ByIDPID(userInfo.Sub))
+	if err != nil && !errors.Is(err, sql.ErrNoRows) {
+		return nil, err
+	}
+
+	if user != nil {
+		return user, nil
+	}
+
+	// No mention of it recorded in the DB, so get the internal-apis user ID via remote user
+	remoteUser, err := getRemoteUser(internalAPIHost, oauth2.StaticTokenSource(&oauth2.Token{AccessToken: tokenString}), metaRemoteIP)
+	if err != nil {
+		/*  For now keep the same process, where if we fail to get a remote user, authentication fails.
+		We don't have to check for verified email since IDP already confirmed via OAuth2. This should
+		be different, where we don't need a user id from internal-apis to continue but for now, to
+		keep it simple lets just require it instead of changing primary keys and wallet names. Also as
+		the wallets are populated with the IDP ID these calls will dwindle down to first time users.
+
+		When we do, do it, we should populate IDP_ID for wallet_IDS for all new wallets created, storing
+		all user id based wallet ids as they arise. Then we will be left with old wallets that have not
+		been accessed in a long time and can be wiped. When they sign in at some point in the future
+		they will get the wallet from the wallet sync APIs.
+		*/
+		return nil, err
+	}
+	log.Data["remote_user_id"] = remoteUser.ID
+	log.Debugf("user id retrieved from internal-apis")
+
+	//See if we already have a wallet but under the user ID from internal-apis
+	user, err = GetDBUserG(ByID(remoteUser.ID))
+	if err != nil && !errors.Is(err, sql.ErrNoRows) {
+		return nil, err
+	}
+
+	if user != nil {
+		user.IdpID.SetValid(userInfo.Sub)
+		_, err := user.UpdateG(boil.Infer())
+		if err != nil {
+			return nil, errors.Err(err)
+		}
+		return user, nil
+	}
+
+	// Ok, no user in internal-apis either. Time to create a new one...
+	ctx, cancelFn := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancelFn()
+
+	err = inTx(ctx, storage.Conn.DB.DB, func(tx *sql.Tx) error {
+		localUser, err = getOrCreateLocalUser(tx, models.User{ID: remoteUser.ID, IdpID: null.StringFrom(userInfo.Sub)}, log)
+		if err != nil {
+			return err
+		}
+
+		if localUser.LbrynetServerID.IsZero() {
+			err := assignSDKServerToUser(tx, localUser, rt.LeastLoaded(), log)
+			if err != nil {
+				return err
+			}
+		}
+
+		return nil
+	})
+
+	if err == nil && localUser != nil {
+		currentCache.set(tokenString, localUser)
+	}
+
+	return localUser, err
+}
+
+func checkAuthorization(info *UserInfo) error {
+	audienceRespected := false
+	for _, aud := range info.Aud {
+		if aud == oauthClientID {
+			audienceRespected = true
+		}
+	}
+	if !audienceRespected {
+		return errors.Err("this token is not meant for Odysee APIs")
+	}
+
+	/* If we could get the http request we could valid allowed sources
```

Comment:
What needs to happen for that? Also, looks like there is a typo.

Reply:
Well it would be best if the middleware would just pass the http request to the provider instead of the token. However, it's designed this way, from what I saw to make testing easier. If you had to process a request in the provider, using mock providers would be more costly and I found them being used in tests all over the place. So I did not want to make that sort of change.